File: c#/vs/Common7/IDE/Extensibility Projects/MCpp/2052/Connect.cpp
// 这是主 DLL 文件。

#include "stdafx.h"
#include "Connect.h"

void <%=SAFEOBJNAME%>::Connect::OnAddInsUpdate(System::Array ^%custom)
{
}

void <%=SAFEOBJNAME%>::Connect::OnBeginShutdown(System::Array ^%custom)
{
}

void <%=SAFEOBJNAME%>::Connect::OnConnection(System::Object ^Application, ext_ConnectMode ConnectMode, System::Object ^AddInInst, System::Array ^%custom)
{
    _applicationObject = dynamic_cast<DTE2^>(Application);
    _addInInstance = dynamic_cast<AddIn^>(AddInInst);
    <%BEGIN VSCommand%>if(ConnectMode == ext_ConnectMode::ext_cm_UISetup)
    {
        array<System::Object^>^ contextGUIDs = gcnew array<System::Object^>(0);
        Commands2 ^commands = dynamic_cast<Commands2^>(_applicationObject->Commands);
        String ^toolsMenuName = "Tools";

        //将此命令置于“工具”菜单上。
        //查找 MenuBar 命令栏，该命令栏是容纳所有主菜单项的顶级命令栏:
        _CommandBars ^commandBars = dynamic_cast<CommandBars^>(_applicationObject->CommandBars);
        CommandBar ^menuBarCommandBar = dynamic_cast<CommandBar^>(commandBars["MenuBar"]);

        //在 MenuBar 命令栏上查找“工具”命令栏:
        CommandBarControl ^toolsControl = menuBarCommandBar->Controls[toolsMenuName];
        CommandBarPopup ^toolsPopup = dynamic_cast<CommandBarPopup^>(toolsControl);

        //如果希望添加多个由您的外接程序处理的命令，可以重复此 try/catch 块，
        //  只需确保更新 QueryStatus/Exec 方法，使其包含新的命令名。
        try
        {	
            //将一个命令添加到 Commands 集合:
            Command ^command = commands->AddNamedCommand2(_addInInstance, "<%=SAFECOMMANDNAME%>", "<%=SAFEOBJNAME%>", "Executes the command for <%=SAFEOBJNAME%>", true, 59, contextGUIDs, (int)vsCommandStatus::vsCommandStatusSupported+(int)vsCommandStatus::vsCommandStatusEnabled, (int)vsCommandStyle::vsCommandStylePictAndText, vsCommandControlType::vsCommandControlTypeButton);

            //将对应于该命令的控件添加到“工具”菜单:
            if((command) && (toolsPopup))
            {
                command->AddControl(toolsPopup->CommandBar, 1);
            }
        }
        catch(System::ArgumentException ^)
        {
            //如果出现此异常，原因很可能是由于具有该名称的命令
            //  已存在。如果确实如此，则无需重新创建此命令，并且
            //  可以放心忽略此异常。
        }
    }<%END VSCommand%>
}

void <%=SAFEOBJNAME%>::Connect::OnStartupComplete(System::Array ^%custom)
{
}

void <%=SAFEOBJNAME%>::Connect::OnDisconnection(ext_DisconnectMode removeMode, System::Array ^%custom)
{
}

<%BEGIN VSCommand%>
void <%=SAFEOBJNAME%>::Connect::Exec(String ^CmdName, vsCommandExecOption ExecuteOption, Object ^%VariantIn, Object ^%VariantOut, bool %handled)
{
    handled = false;
    if(ExecuteOption == vsCommandExecOption::vsCommandExecOptionDoDefault)
    {
        if(!CmdName->CompareTo("<%=SAFEOBJNAME%>.Connect.<%=SAFECOMMANDNAME%>"))
        {
            handled = true;
            return;
        }
    }
}

void <%=SAFEOBJNAME%>::Connect::QueryStatus(String ^CmdName, vsCommandStatusTextWanted NeededText, vsCommandStatus %StatusOption, Object ^%CommandText)
{
    if(NeededText == vsCommandStatusTextWanted::vsCommandStatusTextWantedNone)
    {
        if(!CmdName->CompareTo("<%=SAFEOBJNAME%>.Connect.<%=SAFECOMMANDNAME%>"))
        {
            StatusOption = (vsCommandStatus)(vsCommandStatus::vsCommandStatusSupported+vsCommandStatus::vsCommandStatusEnabled);
            return;
        }
    }
}<%END VSCommand%>

File: c#/vs/Common7/IDE/Extensibility Projects/VBasic/Connect.vb
Imports Microsoft.VisualStudio.CommandBars
Imports Extensibility
Imports EnvDTE
Imports EnvDTE80

Public Class Connect
	
    Implements IDTExtensibility2
	<%BEGIN VSCommand%>Implements IDTCommandTarget<%END VSCommand%>

    Private _applicationObject As DTE2
    Private _addInInstance As AddIn

    ''' <summary>
    ''' Implements the constructor for the Add-in object. Place your initialization code within this method.
    ''' </summary>
    Public Sub New()

    End Sub

    ''' <summary>
    ''' Implements the OnConnection method of the IDTExtensibility2 interface. 
    ''' Receives notification that the Add-in is being loaded.
    ''' </summary>
    ''' <param name='application'>Root object of the host application.</param>
    ''' <param name='connectMode'>Describes how the Add-in is being loaded.</param>
    ''' <param name='addInInst'>Object representing this Add-in.</param>
    ''' <remarks></remarks>
    Public Sub OnConnection(ByVal application As Object, ByVal connectMode As ext_ConnectMode, ByVal addInInst As Object, ByRef custom As Array) Implements IDTExtensibility2.OnConnection
        _applicationObject = CType(application, DTE2)
        _addInInstance = CType(addInInst, AddIn)
        <%BEGIN VSCommand%>If connectMode = ext_ConnectMode.ext_cm_UISetup Then

            Dim commands = CType(_applicationObject.Commands, Commands2)
            Dim toolsMenuName = "Tools"

            'Place the command on the tools menu.
            'Find the MenuBar command bar, which is the top-level command bar holding all the main menu items:
            Dim commandBars = CType(_applicationObject.CommandBars, CommandBars)
            Dim menuBarCommandBar As CommandBar = commandBars.Item("MenuBar")

            'Find the Tools command bar on the MenuBar command bar:
            Dim toolsControl As CommandBarControl = menuBarCommandBar.Controls.Item(toolsMenuName)
            Dim toolsPopup = CType(toolsControl, CommandBarPopup)

            Try
                'Add a command to the Commands collection:
                Dim command = commands.AddNamedCommand2(_addInInstance, "<%=SAFECOMMANDNAME%>", "<%=SAFEOBJNAME%>",
                                                        "Executes the command for <%=SAFEOBJNAME%>", True, 59, Nothing,
                                                        vsCommandStatus.vsCommandStatusSupported Or vsCommandStatus.vsCommandStatusEnabled,
                                                        vsCommandStyle.vsCommandStylePictAndText, vsCommandControlType.vsCommandControlTypeButton)

                'Find the appropriate command bar on the MenuBar command bar:
                command.AddControl(toolsPopup.CommandBar, 1)
            Catch argumentException As ArgumentException
                'If we are here, then the exception is probably because a command with that name
                '  already exists. If so there is no need to recreate the command and we can 
                '  safely ignore the exception.
            End Try

        End If<%END VSCommand%>
    End Sub

    ''' <summary>
    ''' Implements the OnDisconnection method of the IDTExtensibility2 interface. 
    ''' Receives notification that the Add-in is being unloaded.
    ''' </summary>
    ''' <param name='disconnectMode'>Describes how the Add-in is being unloaded.</param>
    ''' <param name='custom'>Array of parameters that are host application specific.</param>
    ''' <remarks></remarks>
    Public Sub OnDisconnection(ByVal disconnectMode As ext_DisconnectMode, ByRef custom As Array) Implements IDTExtensibility2.OnDisconnection
    End Sub

    ''' <summary>
    ''' Implements the OnAddInsUpdate method of the IDTExtensibility2 interface. 
    ''' Receives notification that the collection of Add-ins has changed.
    ''' </summary>
    ''' <param name='custom'>Array of parameters that are host application specific.</param>
    ''' <remarks></remarks>
    Public Sub OnAddInsUpdate(ByRef custom As Array) Implements IDTExtensibility2.OnAddInsUpdate
    End Sub

    ''' <summary>
    ''' Implements the OnStartupComplete method of the IDTExtensibility2 interface. 
    ''' Receives notification that the host application has completed loading.
    ''' </summary>
    ''' <param name='custom'>Array of parameters that are host application specific.</param>
    ''' <remarks></remarks>
    Public Sub OnStartupComplete(ByRef custom As Array) Implements IDTExtensibility2.OnStartupComplete
    End Sub

    ''' <summary>
    ''' Implements the OnBeginShutdown method of the IDTExtensibility2 interface. 
    ''' Receives notification that the host application is being unloaded.
    ''' </summary>
    ''' <param name='custom'>Array of parameters that are host application specific.</param>
    ''' <remarks></remarks>
    Public Sub OnBeginShutdown(ByRef custom As Array) Implements IDTExtensibility2.OnBeginShutdown
    End Sub
	<%BEGIN VSCommand%>
    ''' <summary>
    ''' Implements the QueryStatus method of the IDTCommandTarget interface. 
    ''' This is called when the command's availability is updated
    ''' </summary>
    ''' <param name='commandName'>The name of the command to determine state for.</param>
    ''' <param name='neededText'>Text that is needed for the command.</param>
    ''' <param name='status'>The state of the command in the user interface.</param>
    ''' <param name='commandText'>Text requested by the neededText parameter.</param>
    ''' <remarks></remarks>
    Public Sub QueryStatus(ByVal commandName As String, ByVal neededText As vsCommandStatusTextWanted, ByRef status As vsCommandStatus, ByRef commandText As Object) Implements IDTCommandTarget.QueryStatus
        If neededText = vsCommandStatusTextWanted.vsCommandStatusTextWantedNone Then
            If commandName = "<%=SAFEOBJNAME%>.Connect.<%=SAFECOMMANDNAME%>" Then
                status = vsCommandStatus.vsCommandStatusEnabled Or vsCommandStatus.vsCommandStatusSupported
            Else
                status = vsCommandStatus.vsCommandStatusUnsupported
            End If
        End If
    End Sub

    ''' <summary>
    ''' Implements the Exec method of the IDTCommandTarget interface.
    ''' This is called when the command is invoked.
    ''' </summary>
    ''' <param name='commandName'>The name of the command to execute.</param>
    ''' <param name='executeOption'>Describes how the command should be run.</param>
    ''' <param name='varIn'>Parameters passed from the caller to the command handler.</param>
    ''' <param name='varOut'>Parameters passed from the command handler to the caller.</param>
    ''' <param name='handled'>Informs the caller if the command was handled or not.</param>
    ''' <remarks></remarks>
    Public Sub Exec(ByVal commandName As String, ByVal executeOption As vsCommandExecOption, ByRef varIn As Object, ByRef varOut As Object, ByRef handled As Boolean) Implements IDTCommandTarget.Exec
        handled = False
        If executeOption = vsCommandExecOption.vsCommandExecOptionDoDefault Then
            If commandName = "<%=SAFEOBJNAME%>.Connect.<%=SAFECOMMANDNAME%>" Then
                handled = True
                Exit Sub
            End If
        End If
    End Sub<%END VSCommand%>
End Class
